### ISSUE TITLE
Footgun: Storing `Record` type objects within crates.

### ISSUE THREAD

**Issue by Neohertz** [Neohertz/crate#8]
Crate's API was originally designed to be used with primitives and arrays, but there may be times where you'd like to store a `Record` type within crate. 

Crate can be used this way, but It's important to understand that this **will** lead to unintended behavior if you aren't careful.

## Example
Let's say we have the following crate structure and type.
```ts
interface CrateType {
  record: Record<string, string>
}

const myCrate = new Crate({
  record: {
    name: "Paul",
  }
})
```
Now, let's try to update `record.name`. This would work how you'd expect it to.
```ts
print(myCrate.getState((s) => s.record)) // ✅ { name: "Paul" }

myCrate.update({
  record: { name: "John", favoriteFood: "Pie" }
}).expect()

print(myCrate.getState((s) => s.record)) // ✅ { name: "John", favoriteFood: "Pie" }
```
Now, let's try to "reset" the object to `{}` and read the state.
```ts
myCrate.update({
  record: {}
})

print(myCrate.getState((s) => s.record)) // ❌ { name: "John", favoriteFood: "Pie" }
```

As you can see, our state did not update like we expected it to.

## So, what happened?
Let's take a look at how crate handles updates.

```ts
myCrate.update(
  { // 1
    a:  { // 2
      nested: { // 3 
        value: 10 
      }
    } 
  }
)
```
When crate receives an update call, it will recursively scan the object for changes. In the above example, crate do the following steps. (simplified)

```bash
found object 1 -> iterate through 1 
  -> found object 2 -> iterate through object 2 
    -> found object 3 -> iterate through object 3
      -> found modification (key: value)
           | set root.a.nested["value"] = 10
```
When crate receives an update call, crate recursively scans each object within the request. If it finds a key that doesn't point to another object, it will interpret value as mutations to the original state. This is how you are able to update deeply nested primitives within crate selectively, like so:
```ts
const myCrate = new Crate({
 a: { nested: { value: 10 } },
 value: 50
})
myCrate.update({ a: { nested: { value: 10 } } })
```

Lets take the original example. Given this update call, crate would take the steps below.
```ts
interface CrateType {
  record: Record<string, string>
}

const myCrate = new Crate(
  { // 1
    record: {} // 2
  }
)

myCrate.update({ record: { name: "Neohertz" } })
```

```bash
found object 1 -> iterate through 1 
  -> found object 2 -> iterate through object 2 
    -> found modification (key: name)
       | set root.record["name"] = 10
```
As you can see, crate treats this object as a dictionary of mutations, not an object to replace it with.

>TLDR: when you pass `{}` as a value you're essentially telling crate that no modifications have occurred. Furthermore, updating Records like this **will only ever add or mutate keys in your record.**

## Okay, what's the fix?

> [!NOTE]
> Ideally, you should avoid this pattern entirely and use an array for dynamic data.

To keep this behavior predicable and easy to understand, **all updates to objects who's keys are undefined should use a transformer method.** This is because transformers will **force a state update when they return an object**.

Following our above example, let's give this pattern a try:

```ts
print(myCrate.getState((s) => s.record)) // ✅ { name: "John", favoriteFood: "Pie" }

myCrate.update({
  record: () => ({})
})

print(myCrate.getState((s) => s.record)) // ✅  {}
```

As you can see, it works exactly how we expect it to.

## Protections `v1.0.3-rc.1`
- Crate's update function type has been updated to throw a type error if you attempt to update a record type with an object instead of a transformer (in most cases).
- Crate will also throw an error message if you attempt to set said value to an empty object.
